test/raft/replication.hh

```diff
@@ -888,10 +896,32 @@ template <typename Clock>
 future<> raft_cluster<Clock>::add_entries(size_t n, std::optional<size_t> server) {
     size_t end = _next_val + n;
     while (_next_val != end) {
+        co_await add_entry(_next_val, server);
+        _next_val++;
+    }
+}
+
+// Add consecutive integer entries to a leader concurrently
+template <typename Clock>
+future<> raft_cluster<Clock>::add_entries_concurrent(size_t n, std::optional<size_t> server) {
+    size_t end = _next_val + n;
+    std::vector<future<>> futures;
+    futures.reserve(n);
+    while (_next_val != end) {
+        futures.push_back(add_entry(_next_val++, server));
+    }
+    for (auto& f: futures) {
+        co_await std::move(f);
```

Comment:
This is `parallel_for_each`. So the test was broken by 1ab262e86ba88c19b157c33da0c7347a449c59ab when we disabled parallel command application because it makes snapshot non deterministic. Does it pass for you in debug mode?

Reply:
added optional commutative hashes

Reply:
This is not enough. Adding commutative state machine is how we tried to fix it last time as well, but then the test started failing on snapshot verification. The snapshot verification checks that if the latest value the snapshot includes is X then the hash value also includes all numbers up to X, but with re-ordering it may not hold. For example if a snapshot's latest value is 5 the hash at this point may include only 1, 2, 3, 5 and 4 is still in the log because it was reordered with 5.

Reply:
Is there something preventing `_output.snp `in `fsm::apply_snapshot` to be overwritten?

* we approach memory limit, so `applier_fiber` runs snapshots on each entry
* it gets scheduled two times one after the other, applying two commands and taking snapshots for each of them
* when `io_fiber` eventually runs it sees only the last snapshot in `batch.snp`, so the previous one is 'leaked'